File: frontend/src/lib/api/knowledge.ts
/**
 * API client for Knowledge Sources — source catalog and connection CRUD.
 */

import { apiFetch } from "./client";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface TransportInfo {
  transport: string;
  label: string;
  auth_methods: string[];
  auth_fields: Record<string, string[]>;
}

export interface KnowledgeSourceInfo {
  key: string;
  label: string;
  icon: string;
  description: string;
  capabilities: string[];
  transports: TransportInfo[];
  available: boolean;
  connection_count: number;
}

export interface ConnectionInfo {
  id: string;
  source_type: string;
  name: string;
  transport: string;
  auth_method: string;
  config: Record<string, unknown>;
  scope: Record<string, unknown>;
  enabled: boolean;
  credentials_configured: boolean;
  status: string;
  status_detail: string | null;
  last_sync_at: string | null;
  last_success_at: string | null;
  latency_ms: number | null;
  created_at: string | null;
  updated_at: string | null;
}

export interface KnowledgeOverview {
  sources: KnowledgeSourceInfo[];
  connections: ConnectionInfo[];
}

export interface ConnectionCreateRequest {
  source_type: string;
  name: string;
  transport: string;
  auth_method: string;
  config?: Record<string, unknown>;
  credentials?: Record<string, unknown>;
  scope?: Record<string, unknown>;
}

export interface ConnectionUpdateRequest {
  name?: string;
  transport?: string;
  auth_method?: string;
  config?: Record<string, unknown>;
  credentials?: Record<string, unknown>;
  scope?: Record<string, unknown>;
  enabled?: boolean;
}

export interface ConnectionHealthResponse {
  id: string;
  status: string;
  status_detail: string | null;
  latency_ms: number | null;
}

// ---------------------------------------------------------------------------
// API calls
// ---------------------------------------------------------------------------

export function getKnowledgeOverview(token: string): Promise<KnowledgeOverview> {
  return apiFetch<KnowledgeOverview>("/knowledge/overview", { token });
}

export function listKnowledgeSources(token: string): Promise<KnowledgeSourceInfo[]> {
  return apiFetch<KnowledgeSourceInfo[]>("/knowledge/sources", { token });
}

export function listConnections(
  token: string,
  sourceType?: string,
): Promise<ConnectionInfo[]> {
  const query = sourceType ? `?source_type=${encodeURIComponent(sourceType)}` : "";
  return apiFetch<ConnectionInfo[]>(`/knowledge/connections${query}`, { token });
}

export function createConnection(
  token: string,
  body: ConnectionCreateRequest,
): Promise<ConnectionInfo> {
  return apiFetch<ConnectionInfo>("/knowledge/connections", {
    method: "POST",
    token,
    body,
  });
}

export function updateConnection(
  token: string,
  connectionId: string,
  body: ConnectionUpdateRequest,
): Promise<ConnectionInfo> {
  return apiFetch<ConnectionInfo>(`/knowledge/connections/${connectionId}`, {
    method: "PUT",
    token,
    body,
  });
}

export function deleteConnection(token: string, connectionId: string): Promise<void> {
  return apiFetch<void>(`/knowledge/connections/${connectionId}`, {
    method: "DELETE",
    token,
  });
}

export function checkConnectionHealth(
  token: string,
  connectionId: string,
): Promise<ConnectionHealthResponse> {
  return apiFetch<ConnectionHealthResponse>(
    `/knowledge/connections/${connectionId}/health`,
    { method: "POST", token, body: {} },
  );
}

File: backend/app/schemas/knowledge.py
"""DTOs for the Knowledge Sources API.

Security invariant: credentials are write-only — never returned in responses.
The boolean `credentials_configured` signals presence without exposing values.
"""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


# ---------------------------------------------------------------------------
# Source catalog (static, from registry)
# ---------------------------------------------------------------------------


class TransportInfo(BaseModel):
    transport: str
    label: str
    auth_methods: list[str] = Field(default_factory=list)
    auth_fields: dict[str, list[str]] = Field(default_factory=dict)


class KnowledgeSourceInfo(BaseModel):
    """A source type as seen by the UI — rendered generically."""

    key: str
    label: str
    icon: str
    description: str
    capabilities: list[str] = Field(default_factory=list)
    transports: list[TransportInfo] = Field(default_factory=list)
    available: bool = True
    connection_count: int = 0


# ---------------------------------------------------------------------------
# Connection (instance)
# ---------------------------------------------------------------------------


class ConnectionInfo(BaseModel):
    """One configured connection — the UI renders this per-instance."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    source_type: str
    name: str
    transport: str
    auth_method: str
    config: dict = Field(default_factory=dict)
    scope: dict = Field(default_factory=dict)
    enabled: bool = True
    credentials_configured: bool = False

    # Health
    status: str = "unknown"
    status_detail: str | None = None
    last_sync_at: datetime | None = None
    last_success_at: datetime | None = None
    latency_ms: int | None = None

    created_at: datetime | None = None
    updated_at: datetime | None = None


class ConnectionCreateRequest(BaseModel):
    """Create a new connection to a knowledge source."""

    source_type: str
    name: str
    transport: str
    auth_method: str
    config: dict = Field(default_factory=dict)
    credentials: dict = Field(default_factory=dict)  # write-only
    scope: dict = Field(default_factory=dict)


class ConnectionUpdateRequest(BaseModel):
    """Update an existing connection. Omitted fields are unchanged."""

    model_config = ConfigDict(extra="ignore")

    name: str | None = None
    transport: str | None = None
    auth_method: str | None = None
    config: dict | None = None
    credentials: dict | None = None  # write-only; omit to keep current
    scope: dict | None = None
    enabled: bool | None = None


class ConnectionHealthResponse(BaseModel):
    id: uuid.UUID
    status: str
    status_detail: str | None = None
    latency_ms: int | None = None


# ---------------------------------------------------------------------------
# Overview (one-shot load for the UI)
# ---------------------------------------------------------------------------


class KnowledgeOverview(BaseModel):
    """Everything the Knowledge Sources section needs in one round trip."""

    sources: list[KnowledgeSourceInfo]
    connections: list[ConnectionInfo]
